File: worker/tests/fixtures/source/python/valid.py
"""Valid Python fixture: known structure for parse and resolution pass tests.

Expected parse output:
  - FILE node: file:worker/tests/fixtures/source/python/valid.py
  - FUNCTION nodes: authenticate_user, get_user_by_id, hash_password, _validate_token
  - CLASS node: UserService
  - PARAMETER nodes on each function
"""
from __future__ import annotations

import hashlib


class UserService:
    def __init__(self, db):
        self.db = db

    def authenticate_user(self, username: str, password: str) -> dict | None:
        hashed = hash_password(password)
        return self.db.execute(
            "SELECT * FROM users WHERE username = ? AND password_hash = ?",
            (username, hashed),
        ).fetchone()

    def get_user_by_id(self, user_id: int) -> dict | None:
        return self.db.execute(
            "SELECT id, username, email FROM users WHERE id = ?", (user_id,)
        ).fetchone()


def hash_password(password: str) -> str:
    return hashlib.sha256(password.encode()).hexdigest()


def _validate_token(token: str) -> bool:
    return isinstance(token, str) and len(token) == 64

File: worker/tests/fixtures/source/typescript/valid.ts
/**
 * Valid TypeScript fixture: known structure for parse and resolution pass tests.
 *
 * Expected parse output:
 *   - FILE node
 *   - FUNCTION nodes: validateJWT, createUser, hashPassword, _extractClaims
 *   - CLASS node: AuthService
 *   - PARAMETER nodes on each function
 */

import * as crypto from "crypto";

export class AuthService {
  constructor(private readonly db: { query: (sql: string, params: unknown[]) => Promise<unknown[]> }) {}

  async validateJWT(token: string): Promise<{ userId: string } | null> {
    const claims = _extractClaims(token);
    if (!claims) return null;
    return { userId: claims.sub };
  }

  async createUser(email: string, password: string): Promise<string> {
    const hashed = hashPassword(password);
    const rows = await this.db.query(
      "INSERT INTO users (email, password_hash) VALUES (?, ?) RETURNING id",
      [email, hashed]
    );
    return (rows[0] as { id: string }).id;
  }
}

export function hashPassword(password: string): string {
  return crypto.createHash("sha256").update(password).digest("hex");
}

function _extractClaims(token: string): { sub: string } | null {
  try {
    const [, payload] = token.split(".");
    return JSON.parse(Buffer.from(payload, "base64url").toString("utf8")) as { sub: string };
  } catch {
    return null;
  }
}
